safenode/src/node/replication.rs

```diff
@@ -0,0 +1,155 @@
+        info!("Churn detected, the new_peers in our closest_group is: {new_peers:?}");
+        // send out our data address to them
+        // todo: should do the same for reg storage
+        let data_addresses: Vec<DataAddress> = chunk_storage
+            .addrs()
+            .into_iter()
+            .map(DataAddress::Chunk)
+            .collect();
+        for peer in new_peers {
+            let network_clone = network.clone();
+            let data_addresses = data_addresses.clone();
+            // spawn task to send event to each peer
+            let _handle = spawn(async move {
+                let _ = network_clone
+                    .send_request(
+                        Request::Event(Event::ReplicateData(data_addresses.clone())),
+                        peer,
+                    )
+                    .await;
+            });
+        }
+        tokio::time::sleep(Duration::from_secs(20)).await;
+    }
+}
+
+pub(crate) async fn ask_peers_for_data(
+    network: Network,
+    chunk_storage: ChunkStorage,
+    register_storage: RegisterStorage,
+    data_addresses: Vec<DataAddress>,
+) {
+    // for each addr, spawn a task to request for the data corresponding to the addr from its
+    // closest peers and store them.
+    for addr in data_addresses {
+        let req = match addr {
+            DataAddress::Chunk(addr) => {
+                if chunk_storage.get(&addr).await.is_ok() {
+                    continue;
+                }
+                Request::Data(DataRequest::Query(Query::GetChunk(addr)))
+            }
+            DataAddress::Register(addr) => {
+                if register_storage.addrs().await.contains(&addr) {
+                    continue;
+                }
+                Request::Data(DataRequest::Query(Query::Register(RegisterQuery::Get(
+                    addr,
+                ))))
+            }
+            DataAddress::Spend(_) => todo!(),
+        };
+        let network = network.clone();
+        let chunk_storage = chunk_storage.clone();
+        let _register_storage = register_storage.clone();
+        let _handle = spawn(async move {
+            if let Ok(responses) = network.node_send_to_closest(&req).await {
+                for resp in responses.iter().flatten() {
+                    if let Response::Data(DataResponse::Query(QueryResponse::GetChunk(Ok(chunk)))) =
+                        resp
+                    {
+                        if chunk_storage.store(chunk).await.is_ok() {
+                            break;
+                        }
+                    } else if let Response::Data(DataResponse::Query(QueryResponse::GetRegister(
+                        Ok(_register),
+                    ))) = resp
+                    {
+                        // todo
+                        // register_storage.update(register).await.ok();
+                    };
+                }
+            }
+        });
+    }
+}
+
+async fn check_for_group_change(
```

Comment:
we don't need to carry out an explicit periodic check.
just carry out replication on `node add` & `dead node` shall be enough